utils/src/vec.rs

```diff
@@ -0,0 +1,848 @@
+//! A vector type that guarantees at least one element.
+
+use crate::TryFromIterator;
+#[cfg(not(feature = "std"))]
+use alloc::{vec, vec::Vec};
+use bytes::{Buf, BufMut};
+use commonware_codec::{EncodeSize, RangeCfg, Read, Write};
+use core::{
+    num::NonZeroUsize,
+    ops::{Deref, DerefMut, Index, IndexMut},
+    slice::SliceIndex,
+};
+use thiserror::Error;
+
+#[cfg(not(feature = "std"))]
+type VecIntoIter<T> = alloc::vec::IntoIter<T>;
+#[cfg(feature = "std")]
+type VecIntoIter<T> = std::vec::IntoIter<T>;
+
+/// Errors that can occur when creating a [`NonEmptyVec`].
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// The collection was empty.
+    #[error("cannot create NonEmptyVec from empty collection")]
+    Empty,
+}
+
+/// A vector that is guaranteed to contain at least one element.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct NonEmptyVec<T>(Vec<T>);
+
+impl<T> NonEmptyVec<T> {
+    /// Creates a new [`NonEmptyVec`] with a single element.
+    pub fn new(first: T) -> Self {
+        Self(vec![first])
+    }
+
+    /// Creates a [`NonEmptyVec`] from a [`Vec`] without checking if it's empty.
+    ///
+    /// # Panics
+    ///
+    /// Methods on the returned `NonEmptyVec` will panic if the vector is empty.
+    pub const fn from_unchecked(vec: Vec<T>) -> Self {
```

Comment:
I feel like it would be better to just add an assertion here?

Reply:
But yeah, I agree should assert here.